document.py

```diff
@@ -13,60 +16,119 @@ def __init__(self, sizeX: int, sizeY: int) -> None:
+    def applyActionToAgent(self, agent: Agent, action: Union[Action, None]) -> bool:
         inGridAgent = self.grid[agent.x][agent.y]
         assert(inGridAgent == agent)
 
-        ##############################################
-        # TEMP
-        Xmove = randrange(-1,2,1)
-        Ymove = randrange(-1,2,1)
-        XNew = agent.x + Xmove
-        YNew = agent.y + Ymove
-
-        if(XNew < 0): XNew = 0
-        if(XNew >= self.size[0]): XNew = self.size[0]-1
-        if(YNew < 0): YNew = 0
-        if(YNew >= self.size[1]): YNew = self.size[1]-1
-        self.grid[agent.x][agent.y] = None
-        self.grid[XNew][YNew] = agent
-        agent.x = XNew
-        agent.y = YNew
-
-        agent.SetLastAction(action) # run only if true
-        return True
-        # TEMP
-        ##############################################
+        newPosition = np.array([agent.x, agent.y])
+
+        # current position of grabbed object
+        if agent.grab:
+            targetPosition = newPosition + Action.directionToVector(agent.grab)
+        else:
+            targetPosition = np.array([None, None])
+
+        if action is None:
+            # random movement
+            Xmove = randrange(-1, 2, 1)
+            Ymove = randrange(-1, 2, 1)
+            newPosition += [Xmove, Ymove]
+        else:
+            # carry out action
+            if action.type == Action.release:
+                agent.grab = None
+
+            elif action.type == Action.move:
+                # move to a different square
+                newPosition += Action.directionToVector(action.direction)
 
+            elif action.type == Action.grab:
+                # grab furniture
+                if agent.grab:
+                    # ignore grab actions while holding an object
+                    return False
+
+                targetPosition = newPosition + Action.directionToVector(action.direction)
+
+                # check if there's furniture there
+                grabTarget = self.grid[targetPosition[0]][targetPosition[1]]
+                if self.GetCellCode(targetPosition[0], targetPosition[1]) in [CellCodes.Chair, CellCodes.Table_max, CellCodes.Table_min]:
+                    # TODO: consider legal table-carrying configurations (can you only carry by holding short ends?)
+                    agent.grab = action.direction
+
+            elif action.type == Action.wait:
+                return True
+
+        # check against walking off-scene
+        if newPosition[0] < 0:
+            newPosition[0] = 0
+        if newPosition[0] >= self.size[0]:
+            newPosition[0] = self.size[0]-1
+        if newPosition[1] < 0:
+            newPosition[1] = 0
+        if newPosition[1] >= self.size[1]:
+            newPosition[1] = self.size[1]-1
+
+        # check against walking onto occupied cells - can only push the grabbed object or move onto empty cells
+        if not np.array_equal(newPosition, targetPosition) and \
```

Comment:
What if 2 or more agents want to move to the same empty cell? What if other agents want to move table/chair to targetPosition?

Reply:
Currently, agents perform actions in order and change env state after each one. Later, action validation can be moved to a separate method validating the whole action pool.